File: packages/core/src/parser/__testdata__/sample.r
library(tidyverse)
library(httr2)
source("utils.R")

MAX_RETRIES <- 3
DEFAULT_MODEL <- "claude-opus-4-6"

#' Agent Configuration
#' @export
AgentConfig <- R6::R6Class("AgentConfig",
  public = list(
    model = NULL,
    max_tokens = NULL,
    temperature = NULL,

    initialize = function(model = DEFAULT_MODEL, max_tokens = 4096L, temperature = 0.7) {
      self$model <- model
      self$max_tokens <- max_tokens
      self$temperature <- temperature
    }
  )
)

#' Base Agent Class
#' @export
BaseAgent <- R6::R6Class("BaseAgent",
  public = list(
    name = NULL,
    config = NULL,

    initialize = function(name, config = NULL) {
      self$name <- name
      self$config <- config %||% AgentConfig$new()
      private$.status <- "idle"
    },

    process = function(message) {
      stop("Not implemented")
    },

    get_tools = function() {
      character(0)
    },

    get_status = function() {
      private$.status
    }
  ),

  private = list(
    .status = "idle",

    validate = function(input) {
      stopifnot(is.character(input), nchar(input) > 0)
      TRUE
    }
  )
)

#' Synapse Agent
#' @export
SynapseAgent <- R6::R6Class("SynapseAgent",
  inherit = BaseAgent,
  public = list(
    process = function(message) {
      private$validate(message)
      private$.status <- "active"
      result <- private$call_model(message)
      private$.status <- "idle"
      result
    },

    get_tools = function() {
      c("search", "read", "write")
    }
  ),

  private = list(
    call_model = function(message) {
      # TODO: implement actual API call
      paste("Response to:", message)
    }
  )
)

#' Create an agent
#' @param name Agent name
#' @param config Optional configuration
#' @return SynapseAgent instance
#' @export
create_agent <- function(name, config = NULL) {
  SynapseAgent$new(name, config)
}

load_agents <- function(directory) {
  files <- list.files(directory, pattern = "\\.json$", full.names = TRUE)
  lapply(files, function(f) {
    data <- jsonlite::fromJSON(f)
    create_agent(data$name)
  })
}

# FIXME: add error handling for API calls

File: packages/core/src/parser/__testdata__/sample.lua
local json = require("cjson")
local http = require("socket.http")

local MAX_RETRIES = 3
local DEFAULT_MODEL = "claude-opus-4-6"

---@class AgentConfig
---@field model string
---@field max_tokens number
local AgentConfig = {}
AgentConfig.__index = AgentConfig

function AgentConfig.new(model, max_tokens)
    local self = setmetatable({}, AgentConfig)
    self.model = model or DEFAULT_MODEL
    self.max_tokens = max_tokens or 4096
    return self
end

---@class Agent
local Agent = {}
Agent.__index = Agent

function Agent.new(name, config)
    local self = setmetatable({}, Agent)
    self.name = name
    self.config = config or AgentConfig.new()
    self.status = "idle"
    self._tools = {"search", "read", "write"}
    return self
end

function Agent:process(message)
    assert(type(message) == "string" and #message > 0, "Invalid message")
    self.status = "active"
    local result = self:_call_model(message)
    self.status = "idle"
    return result
end

function Agent:get_tools()
    return self._tools
end

function Agent:_call_model(message)
    -- TODO: implement actual API call
    return "Response to: " .. message
end

local function create_agent(name, config)
    return Agent.new(name, config)
end

local function load_config(path)
    local f = io.open(path, "r")
    if not f then return nil end
    local content = f:read("*a")
    f:close()
    return json.decode(content)
end

-- FIXME: error handling in load_config
return {
    Agent = Agent,
    AgentConfig = AgentConfig,
    create_agent = create_agent,
    MAX_RETRIES = MAX_RETRIES,
}
